File: person-info/cpp/src/person.cpp
#include <format>
#include "person.h"

bwhazel::person::person(std::string name, int birth_year, bwhazel::gender gender) :
    name(name),
    birth_year(birth_year),
    gender(gender)
{}

int bwhazel::person::my_age() const {
    return CURRENT_YEAR - birth_year;
}

std::string bwhazel::person::my_gender() const {
    return static_cast<int>(gender) ? "male" : "female";
}

std::string bwhazel::person::who_am_i() const {
    return std::format("I am {}, {}, and I am {} years old!", name, my_gender(), my_age());
}

File: person-info/as/as-plumbing/person.ts
enum Gender {
    Female = 0,
    Male = 1
}

class Person {
    static currentYear: i32 = 2025;

    name: string;
    birthYear: i32;
    gender: Gender;

    constructor(name: string, birthYear: i32, gender: Gender) {
        this.name = name;
        this.birthYear = birthYear;
        this.gender = gender;
    }

    myAge(): i32 {
        return Person.currentYear - this.birthYear;
    }

    myGender(): string {
        return this.gender > 0 ? 'male' : 'female';
    }

    whoAmI(): string {
        return `I am ${this.name}, ${this.myGender()}, and I am ${this.myAge()} years old!`;
    }
}

export function whoIs(name: string, birthYear: i32, gender: i32): string {
    let genderValue: Gender;
    switch (gender) {
        case 0:
            genderValue = Gender.Female;
            break;
        case 1:
            genderValue = Gender.Male;
            break;
        default:
            throw new Error("Gender is not supported.");
    }

    const person = new Person(name, birthYear, genderValue);
    return person.whoAmI();
}